text/0069-schema-libraries.md

```diff
@@ -0,0 +1,169 @@
+users, either within the same organization or even across organizations.
+
+### 1. Within the same organization
+
+This RFC would allow an organization to define its own libraries of schema
+definitions, which could be reused across many different schemas (say, for
+different webapps owned by the organization).
+For instance, the organization might have common definitions of `User` or
+`Account` that apply in many different applications, and although those
+applications may not want to share entire Cedar schemas, with this RFC they
+could share just the definitions of `User` or `Account`, which could be defined
+once in a central location (in the same or separate libraries).
+
+### 2. Across organizations
+
+For another motivating example, consider identity providers (IdPs) which comply
+with the OpenID Connect standard (OIDC).
+The OIDC standard includes a list of attributes that exist on a user type; this
+is naturally declared as a Cedar entity type.
+With this RFC, anyone could provide a "library" representing Cedar definitions for
+OIDC types, and provide that library as a Cedar schema file at some URL; and then
+other Cedar users could use those definitions simply by importing the file from that URL.
+This would allow the Cedar community to gradually coalesce on the "best" way to
+represent an OIDC user in Cedar.
+
+### Motivations common to both scenarios
+
+In both of the above scenarios (within-organization and cross-organization),
+we obtain three key benefits:
+1. Saving each Cedar user the effort of writing common declarations themselves.
+  This facilitates code reuse in schemas, and makes it easier to get started
+  with Cedar.
+2. Providing a way to define common types and actions in a centralized way,
+  which ensures many schemas agree on the "correct" or "best practices"
+  definitions, and provides a single place to make updates if updates are
+  required.
+3. Facilitating code reuse for Cedar authorization calls, not just schemas.
+  When everyone shares a common definition of `OIDC::User`, the community could
+  conceivably converge on a reusable library function for, e.g., converting an
+  OIDC token into Cedar entity data.
+  This would further make it easier to get started with Cedar.
+  (Note that this RFC does not propose the Cedar team writing or maintaining
+  either library definitions for use in schemas or library functions for use in
+  Cedar authorization calls. It only points out that the community could
+  converge on these things.)
+
+## Detailed design
+
+Import statements are only allowed at the top level, outside of all namespace
+declarations.
+(In the future, another RFC could propose allowing imports in other positions.)
+
+The target of the import must be a raw file containing a valid Cedar schema.
+This schema may contain any definitions that are valid today in Cedar schemas,
+including namespaces, entity type definitions, common type definitions, and
+action declarations.
```

Comment:
Since the actions in a schema can impact authorization, a schema library adding or removing an action from a group could result in unexpected authorization decisions.

Reply:
Would you propose that libraries cannot include action declarations, and can only include entity type definitions and common type definitions?

Reply:
I don't that's worth it. Presumably, if you're using a schema library, you're consuming entity data from the IdP as well. The issue is just that you might not expect a schema to impact authorization. Sub-resource integrity plus documentation should mitigate this concern.